### ISSUE TITLE
"no data" on the screen

### ISSUE THREAD

**Issue by LeenaShekhar** [myaooo/RNNVis#77]
Problem: I cannot see any model names in the dropdown menu. 

1. when I run the command: **python -m rnnvis.main server**, I get the following error:

Static folder: /home/leena/Downloads/RNNVis-master/frontend/dist/static
loading models...
Building model from config/model/rnn.yml
Data retrieved from db.
WARN: No gates tensor available, Are you using RNN?
Failed to load model PTB-RNN

Building model from config/model/gru.yml
Data retrieved from db.
Failed to load model PTB-GRU

Building model from config/model/lstm-small-1.yml
Data retrieved from db.
Failed to load model PTB-LSTM-SMALL

Building model from config/model/lstm.yml
Data retrieved from db.
Failed to load model PTB-LSTM-MEDIUM

Building model from config/model/lstm-large.yml
Data retrieved from db.
Failed to load model PTB-LSTM-LARGE

Building model from config/model/yelp-2-gru-50-1.yml
No dataset with name yelp-2 exists
Failed to load model YELP-2-GRU-50-1

2. I also ran python -m tests.test_language_model --config_path=./config/model/lstm.yml --data_path=./cached_data/simple-examples/data but this also does not create checkpoints. 

What step am I missing? I also do not see any checkpoints in any of the directories.

**Comment by myaooo** [comment]
Hi @LeenaShekhar, have you found a directory `./models` in the project root? If the training is done, there should be checkpoints of the model inside the `./models`.
I would suggest you run `python -m tests.test_procedures --config_path=your_config_file.yml` to train the model.

**Comment by LeenaShekhar** [comment]
Hi @myaooo Thank you for your response. 

There is a model folder under config but that is to keep all the yml files for model configurations. 

I ran this command: python -m tests.test_procedures --config_path=./config/model/lstm-small.yml , but this fails as data_path is None. 

Then I ran:  python -m tests.test_procedures --config_path=./config/model/lstm-small.yml --data_path=./cached_data/simple-examples/data and the training started (without GPU for now) with the following messages:

Not using any gpu devices.
Building model..
Building model from ./config/model/lstm-small.yml
Data retrieved from db.
Preparing data..
data_path: %s ./cached_data/simple-examples/data
Start Training
Start Running Train Graph
Epoch 0:
lr:5.0000
epoch[232/1161] speed:508.4 wps, time: 365.1s > loss: 7.3507, acc-1: 0.0470
epoch[464/1161] speed:516.5 wps, time: 353.6s > loss: 7.0458, acc-1: 0.0480

I was trying to find how frequently it saves the checkpoints (I have not really read the source code as of now), but could not locate that configuration. As of now I do not  see any other model folder. Did I miss any flag or do I just need to wait for some more time for it to save the checkpoint?

**Comment by myaooo** [comment]
It's actually some bugs in the code. Currently, it's not periodically saving checkpoints, it will only save after the training arrives the specified epoch num specified in the model config file.
I am adding the periodical saving in another branch. But it is not finished yet. 
Sorry for the inconvenience. I will try to merge the new dev branch back as soon as possible.

**Comment by LeenaShekhar** [comment]
NP about the bug. 
In this case it will save after 40 epochs since epoch_num: 40 in the yml file? Can you point me to the file then I can modify the code? But if you can do the fix soon then I will wait for it. 

Thank you again for the quick response.

**Comment by LeenaShekhar** [comment]
I was able to train 2 models and evaluate them. When I restarted the server and went to the homepage I can see the model names in the drop down, but I can't see the word or the hidden state clusters. 
See the image below:
![image](https://user-images.githubusercontent.com/12227436/32395984-d95bfed6-c0b9-11e7-8332-c2ef32d5a43b.png)

Log:
python -m rnnvis.main server
Static folder: /home/leena/Downloads/RNNVis-master/frontend/dist/static
loading models...
Building model from config/model/rnn.yml
Data retrieved from db.
WARN: No gates tensor available, Are you using RNN?
Model variables restored from /home/leena/Downloads/RNNVis-master/models/RNN-PTB.
**Model PTB-RNN successfully loaded**

Building model from config/model/gru.yml
Data retrieved from db.
Failed to load model PTB-GRU

Building model from config/model/lstm-small-1.yml
Data retrieved from db.
Failed to load model PTB-LSTM-SMALL

Building model from config/model/lstm.yml
Data retrieved from db.
Model variables restored from /home/leena/Downloads/RNNVis-master/models/LSTM-PTB.
**Model PTB-LSTM-MEDIUM successfully loaded**

Building model from config/model/lstm-large.yml
Data retrieved from db.
Failed to load model PTB-LSTM-LARGE

Building model from config/model/yelp-2-gru-50-1.yml
No dataset with name yelp-2 exists
Failed to load model YELP-2-GRU-50-1

 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [03/Nov/2017 17:04:52] "GET /pos_statistics?model=PTB-RNN&top_k=300 HTTP/1.1" 200 -
127.0.0.1 - - [03/Nov/2017 17:04:54] "OPTIONS /models/evaluate HTTP/1.1" 200 -
understanding hidden states.
['understanding', 'hidden', 'states', '<eos>']
input size: 4
Evaluation done!
([4594, 5140, 600, 2],)
[['understanding', 'hidden', 'states', '<eos>']]
127.0.0.1 - - [03/Nov/2017 17:04:54] "POST /models/evaluate HTTP/1.1" 200 -

Is it not able to load the hidden states properly? I have the mongodb instance running in the background, but I am new to mongodb so don't know if there is a problem in storing the states. How do I check that? 
And usually ends with this error:

Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so.

Check MongoDB status:
service mongod status
● mongod.service - High-performance, schema-free document-oriented database
   Loaded: loaded (/lib/systemd/system/mongod.service; disabled; vendor preset: enabled)
   Active: active (running) since Fri 2017-11-03 00:05:08 EDT; 17h ago
     Docs: https://docs.mongodb.org/manual
 Main PID: 8557 (mongod)
   CGroup: /system.slice/mongod.service
           └─8557 /usr/bin/mongod --config /etc/mongod.conf

Nov 03 00:05:08 leena-Lenovo-ideapad-700-15ISK systemd[1]: Started High-performance, schema-free document-oriented database.

**Comment by myaooo** [comment]
Hi @LeenaShekhar, I have just added the code to save checkpoints after each epoch. You can try to train model without specifying data_path using tests.train_with_db:
```bash
python -m tests.train_with_db --config_path=./config/model/lstm-large3.yml
```

As for the later problem of no clusters, could you please paste the log in the browser console?

**Comment by LeenaShekhar** [comment]
Thank you. I will take an update and run it. 
Currently, when I start the sever and make some selections on the screen I get the following log:

python -m rnnvis.main server
Static folder: /home/leena/Downloads/RNNVis-master/frontend/dist/static
loading models...
Building model from config/model/rnn.yml
Data retrieved from db.
WARN: No gates tensor available, Are you using RNN?
Model variables restored from /home/leena/Downloads/RNNVis-master/models/RNN-PTB.
Model PTB-RNN successfully loaded

Building model from config/model/gru.yml
Data retrieved from db.
Failed to load model PTB-GRU

Building model from config/model/lstm-small-1.yml
Data retrieved from db.
Failed to load model PTB-LSTM-SMALL

Building model from config/model/lstm.yml
Data retrieved from db.
Model variables restored from /home/leena/Downloads/RNNVis-master/models/LSTM-PTB.
Model PTB-LSTM-MEDIUM successfully loaded

Building model from config/model/lstm-large.yml
Data retrieved from db.
Failed to load model PTB-LSTM-LARGE

Building model from config/model/yelp-2-gru-50-1.yml
No dataset with name yelp-2 exists
Failed to load model YELP-2-GRU-50-1

 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [04/Nov/2017 00:27:33] "GET /models/config/PTB-LSTM-MEDIUM HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 00:28:05] "OPTIONS /models/evaluate HTTP/1.1" 200 -
understanding hidden states in a rnn model.
['understanding', 'hidden', 'states', 'in', 'a', 'rnn', 'model', '<eos>']
input size: 8
Evaluation done!
([4594, 5140, 600, 7, 6, 1, 1794, 2],)
[['understanding', 'hidden', 'states', 'in', 'a', '<unk>', 'model', '<eos>']]
127.0.0.1 - - [04/Nov/2017 00:28:05] "POST /models/evaluate HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 00:31:32] "GET /pos_statistics?model=PTB-LSTM-MEDIUM&top_k=300 HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 00:41:58] "OPTIONS /models/evaluate HTTP/1.1" 200 -
understanding rnns
['understanding', '<eos>']
input size: 2
Evaluation done!
([4594, 2],)
[['understanding', '<eos>']]
127.0.0.1 - - [04/Nov/2017 00:41:58] "POST /models/evaluate HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 00:44:12] "GET /pos_statistics?model=PTB-RNN&top_k=300 HTTP/1.1" 200 -

After this I do not get any log or anything on the screen. Isn't is supposed to load the states? 

Sorry for so many questions but I really want to try the application as I think it might be very useful to my work. 

P.S: As far as my previous error was concerned, I think something was wrong in my system. The error is pasted below. 

Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so.

**Comment by myaooo** [comment]
Hi @LeenaShekhar, no problem! I am sorry that my code is not easy to use.

I am wondering could you right click on the interface and click inspect to see the console log in the browser.

Since there is no error throw out from the terminal log, I think it should be the problem in the frontend.

The correct log should be like this

```
Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
/Users/mingyao/workspace/RNNVis-test/frontend/dist/index.html
127.0.0.1 - - [04/Nov/2017 11:02:41] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 11:02:41] "GET /static/css/app.b87fa97229316928b7bd72c9888ca792.css HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 11:02:41] "GET /static/js/manifest.4c2d3c97fc574a82e9b8.js HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 11:02:41] "GET /static/js/vendor.0639498cf33971cc8742.js HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 11:02:41] "GET /static/js/app.8557021d69c3a828d514.js HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 11:02:42] "GET /models/available HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 11:02:43] "GET /favicon.ico HTTP/1.1" 404 -
127.0.0.1 - - [04/Nov/2017 11:02:47] "GET /models/config/PTB-LSTM-SMALL HTTP/1.1" 200 -
127.0.0.1 - - [04/Nov/2017 11:02:49] "GET /co_clusters?model=PTB-LSTM-SMALL&state=state_c&n_cluster=2&top_k=300&mode=raw&layer=0 HTTP/1.1" 200 -
```
The last line of log here, the `GET /co_cluster...` is actually trying to load state data. But actually it will take some time if the data is loaded for the first time (it takes quite long to compute it, but it will be cached after that).

Maybe you should wait for some time, or I can add some log to indicate whether it is computing or not.

**Comment by LeenaShekhar** [comment]
@myaooo  I can see the data on screen now. It took some time so maybe some log would be actually great.
I will play around with it now. Thank you for your help an quick response. Great work. 

I have 1 more question (let me know if I should ask in a new thread):

Is it possible to use it for a custom mode? If yes then what are the steps to follow to use a custom model?

**Comment by myaooo** [comment]
@LeenaShekhar, I think I will need to update the doc for this.

Generally, if you want to customize to use well defined RNN cells that TensorFlow provided, you can modify the `cell_type` in the model config. 

If you want to use customized cells, you may want to look into the `get_gate_tensor` of class `RNNModel` in `./rnnvis/rnn/rnn.py` and add code to support the op name of the state. Not sure whether other bugs might occur.

To use custom data, you may want to look into the `rnnvis/db` and try to add code to convert your data to the needed form. Currently, there are two tasks supported, sentiment analysis and language modeling. The seq2seq is not well tested.

I think the current way of customization in RNNVis has many restrictions. It might need much work to customize, I will try to figure out someway easier to do that. 

If you have any suggestions on what kind of ways that you would like RNNVis to support customizations, feel free to start another issue.

**Comment by LeenaShekhar** [comment]
Thank for for that. I will check the code blocks that you mentioned. Nevertheless it is great. Thank you for closing the ticket.
